### PR TITLE
refactor: secure mongo and add database test helpers

### LINKED ISSUE #521: Secure mongo
The current docker setup runs mongo without any auth. In order to provide better security, copy the config from local-dev and update the functional tests so they do not break.

### PR DESCRIPTION
**Pull request by warrenv**
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

Closes #521

Previously, mongo was being run without any auth.

This commit:
  - Enables mongo auth for a better security profile.
  - Adds database helpers to create and teardown temporary databases
     during testing.
  - Updates the functional tests to use the new database helpers.

To test:
```bash
  $ cp docker-compose.override.yml.example docker-compose.override.yml
  $ docker-compose build
  $ docker-compose up -d
  $ docker-compose exec poet-node test:functional
```

### CODE REVIEW

**Comment by jwicks31** [comment]
I pulled this branch down to review [docker-compose-refresh](https://github.com/poetapp/local-dev/pull/53) and using both the local-dev environment and the node docker-compose file I get 8 failing tests out of 12.

**Comment by warrenv** [comment]
Are you seeing any obvious errors in the output?

**Comment by jwicks31** [comment]
I re-ran the tests inside of both environments. It passed for all 12 this time in the node environment, but when using local dev the same 8 still failed. The reason they are failing is because: 
```
operator: deepEqual
expected: 'string'
actual:   'undefined'
```

They are checking for the `blockHeight`, `blockHash`, `transactionId` in Node A. And the `ipfsFileHash`, `ipfsDirectoryHash`, `blockHeight`, `blockHash`, `transactionId` in Node B.

File: docker-compose.override.yml.example [hunk 1]
```diff
@@ -14,23 +17,23 @@ services:
   # running functional tests. Someone checking out the repo and wanting
   # to just run a poet node only needs bitcoind-1.
   bitcoind-2:
-    # image: ruimarinho/bitcoin-core:0.16.3
+    image: ruimarinho/bitcoin-core:0.16.3
```
Comment (geoffturk) [inline docker-compose.override.yml.example]: 0.17.0

File: mongodb/create_poet_db.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Comment (lautarodragan) [inline mongodb/create_poet_db.sh]: Will this script be only used in tests, or also in deployments for production?

Comment (warrenv) [reply mongodb/create_poet_db.sh]: I'm guessing production would be different - I'm not 100% clear on how production is set up (are we using Atlas?).

Comment (geoffturk) [reply mongodb/create_poet_db.sh]: Yes. We are using Atlas.

Ping @WesleyCharlesBlake regarding the use of the script (in production and QA).

Comment (WesleyCharlesBlake) [reply mongodb/create_poet_db.sh]: this script will not be used on the Atlas mongodb clusters. Atlas allows us to setup users, and specify perms on DBs and collections through its UI, and the DB's get created when we specify the connection URI in the mongo client (ie `mongodb+srv://admin:<PASSWORD>@<some-cluster-id>.mongodb.net/test` will end up creating a `test` db, provided that the credentials provided in the URI have permissions to read/write on that DB.

from a K8s perspectice, we supply the mongo Atlas DB creds in values.yaml, and this creates a config secret in K8s which the node deployment then retrieves to configure mongo auth

File: docker-compose.override.yml.example [hunk 3]
```diff
@@ -14,23 +17,23 @@ services:
...
-    #   -deprecatedrpc=signrawtransaction
-    #   -addnode=bitcoind-1
-    # links:
-    #   - bitcoind-1:bitcoind-1
+    ports:
+      - "19443:18443"
+    command:
+      -printtoconsole
+      -testnet=${USE_BTC_TESTNET:-0}
+     -regtest=${USE_BTC_REGTEST:-1}
```
Comment (krobi64) [inline docker-compose.override.yml.example]: add space

**Comment by krobi64** [comment]
I also can not run `docker exec poet-node test:functional`. It says `Error: No such container: poet-node`.

However I can use `docker-compose exec poet-node bash` and then `npm run test:functional`.

That returns 
```
1..12
# tests 12
# pass  5
# fail  7
```

File: mongodb/create_poet_db.sh (hunk 2 above)
Comment (lautarodragan) [reply mongodb/create_poet_db.sh]: OK. Thanks @geoffturk and @WesleyCharlesBlake.

We should move this file to /tests then. /scripts could work too, if we intend to run it manually.

Comment (warrenv) [reply mongodb/create_poet_db.sh]: I disagree - this is not for tests, it is for bootstrapping a brand new instance of the mongo docker container. It's a hook into how the mongo containers work so you can have an existing set of databases and users already set up when the container starts.

File: docker-compose.override.yml.example (hunk 3 above)
Comment (warrenv) [reply docker-compose.override.yml.example]: done.

**Comment by warrenv** [comment]
@krobi64 

- I updated the instructions to use docker-compose instead of docker.
- The 'out of range errors' are normal - they do not affect the tests.

Sometimes the functional tests fail if they are run immediately after the container start.
Re-running them usually gives a passing result. If not. , it may be necessary to bump the timeout values in the poet-node section of your docker-compose.override.yml:

TIMESTAMP_INTERVAL_IN_SECONDS
BATCH_CREATION_INTERVAL_IN_SECONDS
READ_DIRECTORY_INTERVAL_IN_SECONDS

**Comment by krobi64** [comment]
* [x] Tested changes manually
* [x] Checked accidental architectural/style changes
* [x] Reviewed entire diff
* [ ] Unit tests
* [ ] Documentation
* [x] Filenames and locations

**Pull request merged**